skglm/utils.py

```diff
@@ -1,10 +1,93 @@
+from functools import lru_cache
+
 import numpy as np
+import numba
 from numba import njit
+from numba.experimental import jitclass
+from numba import float32, float64
 
 from numpy.linalg import norm
 from sklearn.utils import check_random_state
 
 
+def spec_to_float32(spec):
+    """Convert a numba specification to an equivalent float32 one.
+
+    Parameters
+    ----------
+    spec : list
+        A list of (name, dtype) for every attribute of a jitclass.
+
+    Returns
+    -------
+    spec32 : list
+        A list of (name, dtype) for every attribute of a jitclass, where float64
+        have been replaced by float32.
+    """
+    spec32 = []
+    for name, dtype in spec:
+        if dtype == float64:
+            dtype32 = float32
+        elif isinstance(dtype, numba.core.types.npytypes.Array):
+            if dtype.dtype == float64:
+                dtype32 = dtype.copy(dtype=float32)
+            else:
+                dtype32 = dtype
+        else:
+            raise ValueError(f"Unknown spec type {dtype}")
+        spec32.append((name, dtype32))
+    return spec32
+
+
+@lru_cache()
+def jit_cached_compile(klass, spec, to_float32=False):
```

Comment:
I am not sure ti understand why we need a separated `jit_cached_compile` function, why not directly doing this in the `compile_clone` function? (I am sure there is a good reason, just asking)

Reply:
because we use lru_cache on this function, but not on the cloning one (that gets fed different potentially different instances every time), while this one receives classes

Reply:
I see, so the key point is the caching of with respect to `klass` and `spec` instead of `instance`

Reply:
yes exactly